From 1b41ef9cceb4c04e809bc8bc5bd072a196fea7e7 Mon Sep 17 00:00:00 2001
From: Li Jun <jun.li@nxp.com>
Date: Wed, 3 Feb 2016 16:06:57 +0800
Subject: [PATCH 0941/5242] MLK-12344-2 usb: chipidea: otg: add vbus
 disconnect for gadget after sleep

commit  218716b50fa1d99757617fdee5361cdfbf0f7ea8 from
https://source.codeaurora.org/external/imx/linux-imx.git

During system sleep, if we switch otg role from gadget to host, and host
vbus is directly controlled by ID signal, we will lose vbus drop event
after resume because the vbus is on both at system suspend and resume, so
we will miss gadget disconnect handling before start host role. This patch
is to fix it by adding gadget disconnect for this case.

Acked-by: Peter Chen <peter.chen@nxp.com>
Signed-off-by: Li Jun <jun.li@nxp.com>
(cherry picked from commit 79aab6fc3c9fca1fa4f83f25a5050ea4afc90f0a)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/usb/chipidea/otg.c |    8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/usb/chipidea/otg.c b/drivers/usb/chipidea/otg.c
index 778dda0..8a9f120 100644
--- a/drivers/usb/chipidea/otg.c
+++ b/drivers/usb/chipidea/otg.c
@@ -230,6 +230,14 @@ void ci_handle_id_switch(struct ci_hdrc *ci)
 			 * external connector status.
 			 */
 			ret = hw_wait_vbus_lower_bsv(ci);
+		else if (ci->vbus_active)
+			/*
+			 * If the role switch happens(e.g. during
+			 * system sleep), and we lose vbus drop
+			 * event, disconnect gadget for it before
+			 * start host.
+			 */
+		       usb_gadget_vbus_disconnect(&ci->gadget);
 
 		ci_role_start(ci, role);
 		/* vbus change may have already occurred */
-- 
1.7.9.5

